src/lib/config.rs

```diff
@@ -46,19 +53,72 @@ pub struct AppConfig {
     pub pub_cert: String,
     pub priv_cert: String,
 }
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct AppFileConfig {
+    pub log_level: LogLevel,
+    pub host: String,
+    pub port: u16,
+    pub pub_cert: String,
+    pub priv_cert: String,
+}
+
 pub fn load_config(service_name: String) -> Result<Config> {
     let args: CliArgument = CliArgument::parse();
 
     println!("Working directory {}", current_dir()?.display());
     println!("Loading config from {}", args.config.display());
+
     let config = std::fs::read_to_string(&args.config)?;
-    let mut config: Config = serde_json::from_str(&config)?;
-    config.app.log_level = args.log_level;
-    config.app.port = args.port;
-    config.app.host = args.host;
-    config.app.name = service_name;
-    config.app.pub_cert = args.pub_cert;
-    config.app.priv_cert = args.priv_cert;
-    println!("App config {:#?}", config.app);
-    Ok(config)
+    let root: serde_json::Value = serde_json::from_str(&config)?;
+
+    let db: DbConfig = serde_json::from_value(
+        root.get("db")
+            .ok_or(ConfigError::DefaultDbConfigNotFoundError(format!(
+                "'db' field not found in config file {}",
+                args.config.display()
+            )))?
+            .to_owned(),
+    )?;
+    let app_file: AppFileConfig = serde_json::from_value(
+        root.get("app")
+            .ok_or(ConfigError::DefaultAppConfigNotFoundError(format!(
+                "'app' field not found in config file {}",
+                args.config.display()
+            )))?
+            .get(&service_name)
+            .ok_or(ConfigError::DefaultAppConfigNotFoundError(format!(
+                "'{}' service field not found in 'app'",
+                service_name
+            )))?
+            .to_owned(),
+    )?;
+
+    let mut app: AppConfig = AppConfig::default();
+    app.log_level = args.log_level.unwrap_or(app_file.log_level);
+    app.port = args.port.unwrap_or(app_file.port);
+    app.host = args.host.unwrap_or(app_file.host);
+    app.pub_cert = args.pub_cert.unwrap_or(app_file.pub_cert);
+    app.priv_cert = args.priv_cert.unwrap_or(app_file.priv_cert);
+    app.name = service_name;
+    println!("App config {:#?}", app);
+
+    Ok(Config { db: db, app: app })
+}
+
+#[derive(Debug)]
+pub enum ConfigError {
+    DefaultDbConfigNotFoundError(String),
+    DefaultAppConfigNotFoundError(String),
```

Comment:
Had to add some match statements to make it work, but now it doesn't have custom errors, what do you think?

Before:
![config_before](https://user-images.githubusercontent.com/29493965/198753326-8928c669-d3b2-42dd-aa90-b20c95d04641.png)

After:
![config_after](https://user-images.githubusercontent.com/29493965/198753333-44fbdbcf-9a2e-4ed5-be39-8c19baa85c49.png)

Reply:
> If we do care about error types(like in a formal library for external audience), use thiserror instead

Do you mean I should change the custom errors in all files to thiserror? I could do that if it sounds better!

Sorry about that btw, I added quite a lot of custom errors because it's the only way I could find to wrap any kind of error in order to use the question mark operator

I tried to prioritize making the code readable and idiomatic, but my lack of experience with Rust might have gotten the best of me :p

Reply:
It's not a must but you'll find out converting to thiserror is fairly simple, saving a lot of code. It's a better way to write custom errors

Reply:
Converted all custom errors to this error, already pushed to local-storage branch

Thanks!